### PR TITLE
fix(awq): stop variable-length calibration collapsing to the last batch

### LINKED ISSUE #3035: [BUG] AWQ calibration silently keeps only the last batch for variable-length inputs (starves MoE experts)
### Describe the bug

`AWQProcessor._layer_input_features()` silently discards all but the most recent calibration batch whenever cached per-module activations have variable sequence lengths.

The collapse logic concatenates cached batches only when every tensor shares the same trailing shape:

- equal shapes → `torch.cat(tensors, dim=0)` — all batches contribute;
- variable lengths (e.g. `[1, 423, H]` + `[1, 36, H]`) → `features[name] = tensors[-1]` — **only the last batch survives**, with no warning and no accounting.

Variable-length batches are the normal case for chat/instruction calibration corpora prepared without padding or packing, so in practice the AWQ scale search (per-channel `x_mean` and the ratio/loss evaluation) runs on a single calibration sample per module while the log still reports the full `nsamples`.

### Impact

For dense models this quietly weakens calibration. For fine-grained MoE models it is much worse: each routed expert only ever sees the tokens routed to it, so collapsing to one batch starves most experts entirely.

Real-world numbers from a 48-layer / 512-expert / top-10 MoE (Qwen3.8-Flash-Next) quantized with a 320-sample, 202,750-token variable-length corpus:

- before: every expert projection calibrated from the final batch only (a few hundred tokens for the whole layer);
- after aggregating all batches: `raw_tokens=202,750` per layer root, with per-expert retained-token accounting — and the resulting checkpoint's per-expert audit changed materially.

### Where

`gptqmodel/looper/awq_processor.py`, `_layer_input_features()` — the `else` branch keeping `tensors[-1]`, and the matching choice documented in `_concat_batch_metadata()`.

### Proposed fix

Since sequence-sensitive replay (attention masks / position ids) genuinely cannot span ragged batches, a safe fix is an **opt-in, model-declared aggregation policy** for pointwise modules (MLP/expert projections): pack ragged captures into deterministic bounded token rows (`[1, retained, hidden]`) so every batch contributes, keep legacy behavior for everything else, and record per-module `raw_tokens` / `retained_tokens` / `mode` stats so the remaining latest-batch fallback is at least visible.

PR with this implementation plus regression tests to follow shortly.

### PR DESCRIPTION
**Pull request by Leonccaa**
## Summary

Fixes #3035.

Variable-length calibration batches cannot be concatenated on dim 0, so `_layer_input_features()` silently kept only `tensors[-1]` per module — every earlier batch was discarded from AWQ scale statistics. For mixture-of-experts models this collapses per-expert calibration to a single batch.

This PR adds a model-declared feature-aggregation policy so de-fused pointwise expert projections pack every ragged batch into deterministic, bounded token rows, plus per-module aggregation accounting. Public model classes with a module tree marked `:moe` inherit the policy; dense-only models remain on the existing path.

## What Changed

- New model hook `awq_input_feature_aggregation(module_name)` returning `{"mode": "token_rows", "capture_root": bool}` or `None`, with an optional explicit `max_tokens` override. `BaseQModel` enables it for the declared expert root and its de-fused pointwise children; `AWQProcessor._feature_aggregation_policy()` validates and applies it.
- New `AWQProcessor._pack_token_rows()`: packs ragged captures into `[1, retained_tokens, hidden]` under a workload-derived bound. The automatic budget uses one largest observed batch equivalent, raised only when needed to retain at least one row from every contributing batch, and never exceeds raw routed rows. Remaining capacity is distributed proportionally to batch length, with rows sampled evenly inside each batch.
- `_record_input_feature()` gains `dedupe_batch`; `record_moe_root_input_feature()` records the shared pre-router expert input when the policy requests `capture_root`.
- Expert lifecycle replay makes one guarded call into the root recorder before expert replay.
- Captures are restored to calibration-batch order before collapse so aggregation and kwargs stay aligned when hooks complete out of order.
- Per-module aggregation stats (`mode`, `raw_tokens`, `retained_tokens`, `batches`) are recorded for both activation and scale features and surfaced in quant log rows. `nsamples` reports retained tokens when a policy is active and is unchanged otherwise.
- Group fallback compares `raw_tokens` (corpus coverage routed to the module) rather than the policy's retained sampling bound. For the unchanged latest-batch fallback, the policy reports the actual retained rows.
- Padding rows are excluded before aggregation, and empty captures do not participate in quota allocation.
- Intentionally out of scope: changing the default variable-length path for models without a policy. Sequence-structured replay cannot span ragged batches without aligned masks and positions, so that path keeps its existing contract and now reports what it retained.

## Review Follow-ups Included

- Reserve one row per non-empty batch so a short batch cannot disappear during sampling.
- Redistribute leftover quota across multiple passes so saturated short batches cannot strand budget.
- Preserve per-sample quant logs while reporting token-row coverage separately.
- Use raw coverage for token-row fallback decisions and actual retained rows for latest-batch decisions.
- Enable the bounded token-row policy for affected public model classes through their declared expert roots.
- Derive each module's token-row budget from its observed calibration batches instead of fixing a model-level token count; retain an explicit policy override for exceptional model requirements.
- Audit every registered public MoE definition and require one unambiguous top-level `:moe` root per module-tree variant. The audit corrected the missing `mlp` marker for GPT-OSS and added Granite MoE Hybrid's Defuser-expanded `block_sparse_moe` expert paths and marker.
- Keep processor-owned feature accounting state concise (`_feature_stats`, `_scale_feature_by_module`, and `_feature_task_names`); the AWQ-prefixed kwargs key remains namespaced because it crosses into model scaling hooks.
- Centralize successful layer finalization so normal completion and both early-exit paths release child/root capture tasks, aggregation telemetry, replay kwargs, module tracking, and thread-local scale context.

## Tests

- [x] Public model-policy coverage for Mixtral, Qwen3 MoE, Qwen3 Next, and MiniMax M2 definitions, including alias-based expert roots and dense-model opt-out.
- [x] Registry-wide expert-root coverage for all 52 detected public MoE model classes/aliases, plus exact pointwise-path expansion checks for the two corrected definitions.
- [x] Variable-length token-row aggregation, padding exclusion, batch ordering, root-capture deduplication, fallback accounting, and bounded-quota property tests.
- [x] Early-completion regression coverage for both an empty scaling configuration and a configuration whose groups are all rejected.
- [x] Combined focused regression selection: `76 passed`.
- [x] End-to-end smoke validation confirmed that all de-fused expert projections report token-row aggregation, bounded retention, raw non-padding coverage, batch counts, and root-versus-direct scale-feature routing.

### CODE REVIEW

File: gptqmodel/looper/awq_processor.py [hunk 1]
```diff
@@ -373,27 +377,104 @@ def _initialize_sample_counts(self) -> None:
...
+        if not policy or not policy.get("capture_root", False):
+            return
+        # Capture only while at least one routed projection under this root has
+        # an active AWQ task. Replay-only passes after task cleanup must not
+        # repopulate the root cache.
+        task_prefix = f"{module_name}."
+        with self.lock:
+            if not any(name.startswith(task_prefix) for name in self.tasks):
+                return
+        self._record_input_feature(module_name, feature, dedupe_batch=True)
```
[the remark this answers is not in the document]
Comment (Leonccaa) [reply gptqmodel/looper/awq_processor.py:502]: The premise doesn't hold for this processor: projection features are captured by `AWQProcessor.pre_process_fwd_hook`, which records `inp[0]` directly — there is no masked hook wrapper in this path. The root capture records the same pre-router `hidden_states` whose per-expert views those projection hooks observe, so the two are consistent by construction; this PR does not introduce an asymmetry.

Masking padded positions out of *both* capture paths would be a reasonable follow-up improvement, but it is orthogonal to this change (and the motivating token_rows use case runs unpadded batch=1 variable-length calibration, where no pad rows exist).

File: gptqmodel/looper/awq_processor.py [hunk 2]
```diff
@@ -827,8 +1027,12 @@ def _should_fallback_group(
             if feat is None or feat.numel() == 0:
                 return True
 
-            hidden = feat.shape[-1] if feat.dim() >= 1 else 1
-            captured_tokens += feat.numel() // max(hidden, 1)
+            feature_stat = getattr(self, "_awq_feature_stats", {}).get(name, {})
+            if "raw_tokens" in feature_stat:
+                captured_tokens += int(feature_stat["raw_tokens"])
```
[the remark this answers is not in the document]
Comment (Leonccaa) [reply gptqmodel/looper/awq_processor.py]: Using `retained_tokens` here would break the feature under common configurations: `should_use_fallback` compares the observed count against absolute thresholds or percent-of-expected-corpus thresholds. With a deliberate sampling bound like `max_tokens=512`, retained is ≤512 for every policy module by design, so any absolute threshold >512 (or any percent threshold of a realistically sized corpus) would force *all* policy modules into fallback permanently.

The question this gate answers is "did enough of the calibration corpus route to this module" — that is `raw_tokens`. `retained_tokens` is the policy's intentional, uniform statistical bound, not evidence of data scarcity. Added a code comment documenting exactly this distinction in ca56974.

File: gptqmodel/looper/awq_processor.py [hunk 3]
```diff
@@ -408,6 +489,60 @@ def _can_concat_batch_tensors(tensors: List[torch.Tensor]) -> bool:
...
+        if retained_tokens == total_tokens:
+            return torch.cat(rows, dim=0).unsqueeze(0), total_tokens
+        if retained_tokens == 1:
+            return rows[0][:1].unsqueeze(0), total_tokens
+
+        sample_indices = (
+            torch.arange(retained_tokens, dtype=torch.int64)
+            * (total_tokens - 1)
+            // (retained_tokens - 1)
+        )
```
[the remark this answers is not in the document]
Comment (Leonccaa) [reply gptqmodel/looper/awq_processor.py]: Correct — thanks. Fixed in ca56974: `_pack_token_rows` now reserves one row per batch and distributes the remaining budget proportionally to batch length (largest remainder, ties by batch order), sampling evenly inside each batch; when `max_tokens` is below the batch count it takes one leading row from evenly spaced batches. Docstring updated to state the exact guarantee, and `test_awq_pack_token_rows_reserves_a_row_from_every_batch` covers the `[5, 1, 5] / max_tokens=4` case (middle batch retained) plus the sub-batch-count budget and determinism.

[the remark this answers is not in the document]
Comment (Leonccaa) [reply gptqmodel/looper/awq_processor.py]: Follow-up in 60507fa: downstream exhaustive enumeration (1,136,541 small-scale combinations) caught an underfill boundary in ca56974 — the single-pass leftover distribution stranded budget when high-remainder batches saturated (lengths [1, 1, 1, 5] with max_tokens=7 returned 6 rows; 3,499 such cases). Length-1 batches are the normal case for MoE experts hit by a single token. Now distributing leftover over multiple passes (capacity always suffices since retained <= total), plus an exhaustive property test asserting retained == min(total, budget) for every combination. Full file: 14 passed.

**Review by Qubitium — CHANGES_REQUESTED** [review]
@Leonccaa  Codex review:

- **P1 — Fallback behavior regressed for legacy ragged inputs.**  
  [`_should_fallback_group`](https://github.com/ModelCloud/GPTQModel/blob/60507fa89fefc4cfa099bf60018eace0ba783da0/gptqmodel/looper/awq_processor.py#L1083-L1090) always uses `raw_tokens`. For `latest_batch`, raw tokens include discarded batches while AWQ uses only the final batch, so the default fallback threshold can be incorrectly bypassed. Use raw counts only for `token_rows`; otherwise use retained rows.

- **P1 — The advertised fix is inactive for all in-tree models.**  
  The patch adds an optional hook lookup, but no public model defines `awq_input_feature_aggregation`. Therefore every current model still follows `latest_batch`. Add model implementations and integration tests, or narrow the PR scope.

- **P1 — Padding tokens can contaminate the new statistics.**  
  [`record_moe_root_input_feature`](https://github.com/ModelCloud/GPTQModel/blob/60507fa89fefc4cfa099bf60018eace0ba783da0/gptqmodel/looper/awq_processor.py#L455-L477) records `hidden_states` without applying the active keep mask. Padded rows can enter token-row sampling and inflate coverage. Apply masking consistently before root capture.

- **P2 — `nsamples` reporting changed for models without a policy.**  
  [`apply_quant`](https://github.com/ModelCloud/GPTQModel/blob/60507fa89fefc4cfa099bf60018eace0ba783da0/gptqmodel/looper/awq_processor.py#L2071-L2084) uses retained rows unconditionally, so legacy `latest_batch` modules now report only the final batch instead of total calibration samples.

- **P2 — Empty captures underfill the token budget.**  
  [`_pack_token_rows`](https://github.com/ModelCloud/GPTQModel/blob/60507fa89fefc4cfa099bf60018eace0ba783da0/gptqmodel/looper/awq_processor.py#L536-L579) fails for zero-row tensors; `[0, 10]` with `max_tokens=1` returns zero rows. Filter empty batches or distribute quotas only across non-empty batches.

Validation: six targeted new unit tests passed; compilation and whitespace checks passed. Full integration testing was blocked by an existing CUDA/CPU device mismatch, and GitHub reports no PR checks.

**Comment by Qubitium** [comment]
@Leonccaa Also I don't see a model or model def that actually enables the new aggregation policy so I cannot run ci tests on this PR with a real (small) model. Can you push a model def change that enables this?

**Comment by Qubitium** [comment]
@Leonccaa Currently doing PR cleanup and refractor. Will let you know when I am done and you can re-test on your end.

**Comment by Qubitium** [comment]
@Leonccaa  I have pushed all the changes.

1. The fix is now automatically applied to all (well-defined) models that has the correct `:moe` tagging in the model def.
2. User do not need to manually set any max_tokens value. It is auto derived now from calirbation data pass. You can override this. 

The rest of small pattern fixes. 

Please test and verify on your end and push any regression you find.

**Review by Qubitium — APPROVED** [review]
Re-reviewed the current head. The earlier requested changes are resolved, and successful AWQ layer finalization now also releases child/root capture tasks, aggregation telemetry, replay kwargs, module tracking, and scale context on both early-completion paths. Focused regression coverage passes with no remaining code findings.

**Comment by Leonccaa** [comment]
> Re-reviewed the current head. The earlier requested changes are resolved, and successful AWQ layer finalization now also releases child/root capture tasks, aggregation telemetry, replay kwargs, module tracking, and scale context on both early-completion paths. Focused regression coverage passes with no remaining code findings.

Thanks for pushing these changes and the final cleanup.

We reviewed the latest head. The automatic :moe enablement and calibration-derived token-row budget address our remaining concerns, including removal of the fixed 512-token cap.

Our focused validation has passed, including a tiny Qwen3-MoE AWQ ExpertsRoutingBypass quantize → save → reload smoke test. We are currently running a full Qwen 3.8-Flash-Next AWQ quantization with this fix. Qwen 3.8-Flash-Next uses ExpertsRoutingBypass, so this run directly exercises the new root-capture path.

We’ll report back after the full quantization and post-quantization checks finish, and will push any regression we find. Thanks again.

**Comment by Qubitium** [comment]
@Leonccaa  e2e testing on real model found more lifecycle bugs in the AWQ processor when this new pr fix/feature is activated. Pending fix incoming.

**Comment by Qubitium** [comment]
@Leonccaa Bug fixed and more refractor passed tests. You can run your own tests now.

**Review by Qubitium — APPROVED** [review]

**Review by Qubitium — APPROVED** [review]

**Comment by Qubitium** [comment]
Validation update for commit `2c1feed4`:

| Check | Result |
|---|---:|
| H100 ARC accuracy (1,172 samples) | 0.5273037542662116 |
| H100 ARC normalized accuracy | 0.552901023890785 |
| Focused AWQ mask tests | 2 passed |
| Broader AWQ/lifecycle/model-policy suite | 127 passed, 1 skipped |

The review findings are covered in the current head: active keep-mask application during root capture, raw-vs-retained fallback accounting, legacy `nsamples` preservation, empty-batch filtering and quota packing, and automatic public `:moe` model policy enablement. The full-model quality gates remain 0.5094 / 0.5486; the ARC values above are the measured q-projection snapshot results.

**Comment by Qubitium** [comment]
@Leonccaa Please create new PR if you find any bugs.
